### ISSUE TITLE
add support for volume_type in block_device of openstack_compute_instance_v2

### ISSUE THREAD

**Issue by rptaylor** [terraform-providers/terraform-provider-openstack#736]
Hello,

According to the  openstack_compute_instance_v2 documentation:
https://www.terraform.io/docs/providers/openstack/r/compute_instance_v2.html#device_type
the block_device block does not support the volume_type option.

However,  openstack_blockstorage_volume_v1 does support volume_type:
https://www.terraform.io/docs/providers/openstack/r/blockstorage_volume_v1.html#volume_type

When implicitly including a volume as part of a compute instance specification, can support for volume_type be added?

As a work around it should be possible to explicitly create a openstack_blockstorage_volume_v1 with volume_type specified, but this requires the additional step of attaching particular volumes to particular instances. It is much easier to use the inline block_device declaration.

### Terraform Version
```
Terraform v0.11.13
+ provider.null v2.1.0
+ provider.openstack v1.17.0
```

### Affected Resource(s)
openstack_compute_instance_v2
openstack_blockstorage_volume_v1 

### Terraform code
```
resource "openstack_compute_instance_v2" "k8s_node" {
  name       = "${var.cluster_name}-k8s-node-${count.index+1}"
  count      = "${var.number_of_k8s_nodes}"
  availability_zone = "${element(var.az_list_kubelet, count.index)}"
  image_name = "${var.image}"
  flavor_id  = "${var.flavor_k8s_node}"
  key_pair   = "${openstack_compute_keypair_v2.k8s.name}"

  network {
    name = "${var.network_name}"
  }

  security_groups = ["${openstack_networking_secgroup_v2.k8s.name}",
    "${openstack_networking_secgroup_v2.bastion.name}",
    "${openstack_networking_secgroup_v2.worker.name}",
    "default",
  ]

  block_device {
    uuid                  = "${var.image}"
    source_type           = "image"
    volume_size           = "${var.node_volume_size_in_gb}"
    boot_index            = 0
    destination_type      = "volume"
    delete_on_termination = true
    volume_type           = "Database"
  }

}
```

### Expected Behavior
The volume should be created with the specified type.

### Actual Behavior
The specified volume_type is ignored.

**Comment by jtopjian** [comment]
> When implicitly including a volume as part of a compute instance specification, can support for volume_type be added?

Yes, it can be added.

It looks like a recent release of OpenStack will allow this: https://docs.openstack.org/nova/latest/user/block-device-mapping.html

> volume_type - Added in microversion 2.67 to the servers create API to support specifying volume type when booting instances.

This support will first need to be added in the Gophercloud library and then added to Terraform. I'll mark this as an upstream enhancement.

For the time being, as a work around, you can do one of these:

* https://www.terraform.io/docs/providers/openstack/r/compute_instance_v2.html#boot-from-an-existing-volume
* https://www.terraform.io/docs/providers/openstack/r/compute_instance_v2.html#boot-instance-and-attach-existing-volume-as-a-block-device

Please let me know if you have any other questions.

**Comment by rptaylor** [comment]
I have been using those workarounds for awhile but soon (due to upstream Kubespray changes) we will need this feature. How can we get it into Gophercloud to start?

**Comment by ozerovandrei** [comment]
@rptaylor Hello,

This feature should be added into the [extensions/bootfromvolume](https://github.com/gophercloud/gophercloud/blob/master/openstack/compute/v2/extensions/bootfromvolume/requests.go#L45) request options.
It will also need some unit tests there: [extensions/bootfromvolume/testing](https://github.com/gophercloud/gophercloud/tree/master/openstack/compute/v2/extensions/bootfromvolume/testing).
And I guess update of acceptance tests will be needed there: [bootfromvolume_test.go](https://github.com/gophercloud/gophercloud/blob/master/acceptance/openstack/compute/v2/bootfromvolume_test.go).

There is also a good [contributor-tutorial](https://github.com/gophercloud/gophercloud/tree/master/docs/contributor-tutorial) to help you get started.

And you'll also need [MICROVERSIONS.md](https://github.com/gophercloud/gophercloud/blob/master/docs/MICROVERSIONS.md) because this feature requires a microversion to be set.

**Comment by jtopjian** [comment]
I opened an issue here (https://github.com/gophercloud/gophercloud/issues/1689) for this. 

@rptaylor This _might_ be something I can work on in the next week.

**Comment by jtopjian** [comment]
https://github.com/gophercloud/gophercloud/issues/1690 added support for this and has been merged. It should be possible to add it to Terraform now.

side note: a microversion wasn't needed in this case since only the request is being altered - not the response.

**Comment by rptaylor** [comment]
Thanks @jtopjian that's fantastic!

https://github.com/gophercloud/gophercloud/pull/1690/files#diff-7e090d1f0240e8edf36d66aad46426e6
This says Compute API microversion 2.67 or later is required, so it should work in Openstack Stein or later according to
https://docs.openstack.org/releasenotes/nova/stein.html

So is adding it to Terraform pretty straightforward, just copy/paste a few lines to pass through an extra attribute? I might be able to take a look if someone can point me in the right direction.

Thanks!

**Comment by jtopjian** [comment]
You'll want to add the new `volume_type` attribute to the schema. But in addition, you'll need to conditionally set a microversion on the client for whenever `volume_type` is not an empty string. For example:

https://github.com/terraform-providers/terraform-provider-openstack/blob/2b58e07318321a06def676f499e8056dc2811f81/openstack/resource_openstack_compute_volume_attach_v2.go#L86-L89

It's important to _only_ set the microversion when `volume_type` is being used, otherwise users who are running an OpenStack release earlier than Stein will see an error message whenever they use the `openstack_compute_instance_v2` resource.

In addition, you'll need to update the vendoring to pull in the changes from Gophercloud:

```
go get github.com/gophercloud/gophercloud/...
go mod tidy
go mod vendor
```

Let us know if you run into any problems.

**Comment by rptaylor** [comment]
@jtopjian can you take a look at #963  ?

**Comment by jtopjian** [comment]
@rptaylor I'm not actively working on this provider any more, but I'll try to find some time to look at this in the next few days.

**Comment by rptaylor** [comment]
PR merged.

**Comment by ozerovandrei** [comment]
Thank you for working on that @rptaylor 
It will be available in `1.26.0` release.
